File: src/main/java/com/example/store/service/impl/BillService.java
package com.example.store.service.impl;

import com.example.store.data.User;
import com.example.store.dto.BillRequest;
import com.example.store.dto.BillResponse;
import com.example.store.service.IBillService;
import com.example.store.service.IUserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
public class BillService implements IBillService {
    public static final String ROLE_EMPLOYEE = "ROLE_EMPLOYEE";
    public static final String ROLE_AFFILIATE = "ROLE_AFFILIATE";
    public static final String ROLE_CUSTOMER = "ROLE_CUSTOMER";
    @Autowired
    private IUserService userService;

    @Override
    public BillResponse generate(List<BillRequest> request) {
        User user = userService.getCurrentUser();

        int discount = 0;
        if (user.getAuthorities().getName().equals(ROLE_EMPLOYEE)) {
            discount = 30;
        } else if (user.getAuthorities().getName().equals(ROLE_AFFILIATE)) {
            discount = 10;
        } else if (user.getAuthorities().getName().equals(ROLE_CUSTOMER)) {
            if(LocalDateTime.now().minusYears(2).isAfter(user.getCreatedDate())){
                discount = 5;
            }
        }

        int itemsTotal = 0;
        for (BillRequest item : request) {
            int itemtotal = item.getPrice() * item.getQty();
            if (!item.getType().equals("groceries") && discount != 0) {
                itemtotal = itemtotal - ((itemtotal * discount) / 100);
            }
            itemsTotal = itemsTotal + itemtotal;
        }

        // Discount of 5 on every 100
        itemsTotal = itemsTotal - (itemsTotal / 100) * 5;

        return BillResponse.builder().netPayableAmount(itemsTotal).build();
    }

}

File: src/main/kotlin/com/example/store/service/impl/BillService.kt
package com.example.store.service.impl

import com.example.store.dto.BillRequest
import com.example.store.dto.BillResponse
import com.example.store.service.IBillService
import com.example.store.service.IUserService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class BillService : IBillService {
    @Autowired
    private lateinit var userService: IUserService

    override fun generate(request: MutableList<BillRequest?>?): BillResponse? {
        val user = userService?.getCurrentUser()
        var discount = 0
        if (user != null) {
            if (user.getAuthorities()?.getName() == ROLE_EMPLOYEE) {
                discount = 30
            } else if (user.getAuthorities()?.getName() == ROLE_AFFILIATE) {
                discount = 10
            } else if (user.getAuthorities()?.getName() == ROLE_CUSTOMER) {
                if (LocalDateTime.now().minusYears(2).isAfter(user.getCreatedDate())) {
                    discount = 5
                }
            }
        }
        var itemsTotal = 0
        if (request != null) {
            for (item in request) {
                var itemtotal = (item?.getPrice())?.times(item?.getQty()!!)
                if (item != null) {
                    if (item.getType() != "groceries" && discount != 0) {
                        if (itemtotal != null) {
                            itemtotal = itemtotal - itemtotal * discount / 100
                        }
                    }
                }
                itemsTotal = itemsTotal + itemtotal!!
            }
        }

        // Discount of 5 on every 100
        itemsTotal = itemsTotal - itemsTotal / 100 * 5
        val billResponse = BillResponse()
        billResponse.setNetPayableAmount(itemsTotal)
        return billResponse
    }

    companion object {
        val ROLE_EMPLOYEE: String? = "ROLE_EMPLOYEE"
        val ROLE_AFFILIATE: String? = "ROLE_AFFILIATE"
        val ROLE_CUSTOMER: String? = "ROLE_CUSTOMER"
    }
}